staging/kos/pkg/apis/network/v1alpha1/types.go

```diff
@@ -19,9 +19,207 @@ package v1alpha1
 import (
 	"time"
 
+	fuzz "github.com/google/gofuzz"
+
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+// ExtendedObjectMeta has extra metadata for an API object.
+// This is maintained by the server, clients can not modify it.
+type ExtendedObjectMeta struct {
+	// Writes identifies the latest write to each part of the object.
+	// +listType=map
+	// +listMapKey=section
+	// +optional
+	Writes WriteSet `json:"writes" protobuf:"bytes,1,name=writes"`
+}
+
+// WriteSet represents a map from section to time
+type WriteSet []ObjectWrite
+
+// ObjectWrite describes a write to part of an object
+type ObjectWrite struct {
+	// Section identifies the part of the object that was written.
+	// Each type of object is broken down into a type-specific set of
+	// sections.
+	Section string `json:"section" protobuf:"bytes,1,name=section"`
+
+	// ServerTime is the time when the write was recorded at the apiserver
+	ServerTime Timestamp `json:"serverTime" protobuf:"bytes,2,name=serverTime"`
+}
+
+// GetServerWriteTime returns the server time of the write to the
+// given section, or zero if there was none.
+func (writes WriteSet) GetServerWriteTime(section string) Timestamp {
+	return writes.GetWrite(section).ServerTime
+}
+
+// GetWrite returns the write to the given section, or zero if there
+// was none.
+func (writes WriteSet) GetWrite(section string) ObjectWrite {
+	for _, wr := range writes {
+		if wr.Section == section {
+			return wr
+		}
+	}
+	return ObjectWrite{}
+}
+
+// SetWrite produces a revised slice that includes the given write.
+// The input is not side-effected.
+func (writes WriteSet) SetWrite(section string, serverTime Timestamp) WriteSet {
+	n := len(writes)
+	var i int
+	for i = 0; i < n && writes[i].Section != section; i++ {
+	}
+	if i == n {
+		return append(WriteSet{{Section: section, ServerTime: serverTime}}, writes...)
+	}
+	return append(append(WriteSet{{Section: section, ServerTime: serverTime}}, writes[:i]...), writes[i+1:]...)
+}
+
+// Diff produces the subset of the given writes that do not overlap
+// with the other writes
+func (writes WriteSet) Diff(others WriteSet) WriteSet {
+	ans := make(WriteSet, 0, len(writes))
+	for _, wr := range writes {
+		owr := others.GetWrite(wr.Section)
+		if owr == (ObjectWrite{}) {
+			ans = append(ans, wr)
+		}
+	}
+	return ans
+}
+
+// UnionLeft produces the union of the receiver and the other writes
+// that do not overlap with the receiver
+func (writes WriteSet) UnionLeft(others WriteSet) WriteSet {
+	ans := append(WriteSet{}, writes...)
+	for _, owr := range others {
+		wr := ans.GetWrite(owr.Section)
+		if wr == (ObjectWrite{}) {
+			ans = append(ans, owr)
+		}
+	}
+	return ans
+}
+
+// UnionMin produces the union of the two write sets, keeping the
+// earlier time for sections written in both sets
+func (writes WriteSet) UnionMin(others WriteSet) WriteSet {
+	ans := others.Diff(writes)
+	for _, wr := range writes {
+		owr := others.GetWrite(wr.Section)
```

Comment:
If `others` does not contain a write for the given section, `GetWrite` returns an empty `ObjectWrite`, which will be added to `ans` with an empty `Section` field. How about adding `owr.Section = wr.Section`?
Same applies to `WriteSet::UnionMax` and the equivalent functions in the internal types file.

Reply:
Yes, this logic is wrong for Min.  For Max, it is guaranteed that the other will not be taken if it is the zero value.

Reply:
>  it is guaranteed that the other will not be taken if it is the zero value

Mmm... how is that guaranteed? AFAIU both `UnionMin` and `UnionMax` set `ServerTime` even if `owr` is the zero value, but `Section` could be left unset.

Reply:
The expectation is that real server times are positive, so will win over zero in Max.
But I am rewriting both to use the extra bool.

Reply:
> The expectation is that real server times are positive, so will win over zero in Max.

But this affects in no way handling of `Section`.